### PR TITLE
Added demo > fuzzy datepicker directive

### PR DESCRIPTION
**Pull request by Blackbaud-AlexKingman**

### CODE REVIEW

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+  </span>
+</p>
+
+<p>
+  <button
+    class="sky-btn sky-btn-secondary"
+    (click)="clearSelectedDates()"
+  >
+    Clear selected dates
+  </button>
```
Comment (blackbaud-johnly) [inline src/demos/datepicker/fuzzy-datepicker-demo.component.html]: These 4 buttons don't have any space between them. I think they should. See https://developer.blackbaud.com/skyux/components/colorpicker#demo for an example.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+  <button
+    class="sky-btn"
+    type="button"
+    (click)="setMinMaxDates()"
+  >
+    Set min and max dates
```
Comment (blackbaud-johnly) [inline src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Should this be a toggle button to turn this setting on and off? That's the behavior I expected based on other demos, but after setting the min/max, I have to refresh the entire page to unset them.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+      Please enter a valid date.
+    </span>
+    <span *ngIf="date.errors.skyFuzzyDate?.yearRequired">
+      Please select a year.
+    </span>
+    <span *ngIf="date.errors.skyFuzzyDate?.futureDisabled">
+      Date cannot be in the future.
+    </span>
+    <span *ngIf="date.errors.skyFuzzyDate?.maxDate">
+      The current date is beyond the maximum allowed date.
```
Comment (blackbaud-johnly) [inline src/demos/datepicker/fuzzy-datepicker-demo.component.html]: I wrote around "minimum" and "maximum" in the property description because that's not really how people talk about dates. I think it's fine to use "min" and "max" in the property names, but I'd suggest rewording to avoid the terminology in our error examples. Also, these error message strike me as too vague. We tell users that dates are too early or too late, but we don't tell them what the thresholds are, so we leave them to guess what the acceptable dates are. The error message shouldn't leave them guessing. What do you think about changing this error message to `This date is earlier than allowed. Please select a date after [min date].`?

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    </span>
+    <span *ngIf="date.errors.skyFuzzyDate?.minDate">
+      The current date is before the minimum allowed date.
```
Comment (blackbaud-johnly) [inline src/demos/datepicker/fuzzy-datepicker-demo.component.html]: As per comments on the previous error message, what do you think about changing this error message to `This date is later than allowed. Please select a date before [max date].`?

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    </span>
+  </p>
+</div>
+
+<p>
+  Selected fuzzy date is {{ selectedDateForDisplay }}
```
Comment (blackbaud-johnly) [inline src/demos/datepicker/fuzzy-datepicker-demo.component.html]: This is currently a sentence fragment with no punctuation at the end. I'd either delete "is" and add a colon after "date:" to make it more like a key-value pair, or else I'd add "The" to the start, lowercase "selected", and add a period to the end to make this a complete sentence.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+      [skyFuzzyDatepickerInput]
+      [skyDatepickerNoValidate]="noValidate"
+      [startingDay]="startingDay"
+      [yearRequired]="yearRequired"
+    />
+  </sky-datepicker>
+</form>
+
+<p>
+  Reactive form selected fuzzy date is {{ reactiveFormSelectedDateForDisplay }}
```
Comment (blackbaud-johnly) [inline src/demos/datepicker/fuzzy-datepicker-demo.component.html]: As with the corresponding text above for the standard fuzzy datepicker, this is currently a sentence fragment with no punctuation at the end. Also, do we really need "Reactive form" at the beginning to distinguish this from the previous text? I don't think so; I think placement on the page makes that clear, I'd delete "Reactive form." And then I'd either capitalize "Selected," delete "is," and add a colon after "date:" to make it more like a key-value pair, or else I'd add "The" to the start and add a period to the end to make this a complete sentence.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    Please select a year.
+  </span>
+  <span *ngIf="reactiveDate.errors.skyFuzzyDate?.futureDisabled">
+    Date cannot be in the future.
+  </span>
+  <span *ngIf="reactiveDate.errors.skyFuzzyDate?.maxDate">
+    The current date is beyond the maximum allowed date.
+  </span>
+  <span *ngIf="reactiveDate.errors.skyFuzzyDate?.minDate">
+    The current date is before the minimum allowed date.
```
Comment (blackbaud-johnly) [inline src/demos/datepicker/fuzzy-datepicker-demo.component.html]: See notes above, but what do you think about changing this error message to `This date is earlier than allowed. Please select a date after [min date].`?

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    Please enter a valid date.
+  </span>
+  <span *ngIf="reactiveDate.errors.skyFuzzyDate?.yearRequired">
+    Please select a year.
+  </span>
+  <span *ngIf="reactiveDate.errors.skyFuzzyDate?.futureDisabled">
+    Date cannot be in the future.
+  </span>
+  <span *ngIf="reactiveDate.errors.skyFuzzyDate?.maxDate">
+    The current date is beyond the maximum allowed date.
```
Comment (blackbaud-johnly) [inline src/demos/datepicker/fuzzy-datepicker-demo.component.html]: See notes above, but what do you think about changing this error message to `This date is later than allowed. Please select a date before [max date].`?

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+  class="sky-error-label"
+>
+  <span *ngIf="reactiveDate.errors.skyFuzzyDate?.invalid">
+    Please enter a valid date.
+  </span>
+  <span *ngIf="reactiveDate.errors.skyFuzzyDate?.yearRequired">
+    Please select a year.
+  </span>
+  <span *ngIf="reactiveDate.errors.skyFuzzyDate?.futureDisabled">
+    Date cannot be in the future.
```
Comment (blackbaud-johnly) [inline src/demos/datepicker/fuzzy-datepicker-demo.component.html]: This is missing "The" or "A" at the beginning, but I'd rewrite this altogether to use active voice: `You cannot specify dates in the future.`

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+<p>
+  Reactive form selected fuzzy date is {{ reactiveFormSelectedDateForDisplay }}
+</p>
+
+<p *ngIf="reactiveDate.errors && (reactiveDate.dirty || reactiveDate.touched)"
+  class="sky-error-label"
+>
+  <span *ngIf="reactiveDate.errors.skyFuzzyDate?.invalid">
+    Please enter a valid date.
```
Comment (blackbaud-johnly) [inline src/demos/datepicker/fuzzy-datepicker-demo.component.html]: I think "enter" should be "select" here because I think we use "select" in all the other messages.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+    Date cannot be in the future.
+  </span>
+  <span *ngIf="reactiveDate.errors.skyFuzzyDate?.maxDate">
+    The current date is beyond the maximum allowed date.
+  </span>
+  <span *ngIf="reactiveDate.errors.skyFuzzyDate?.minDate">
+    The current date is before the minimum allowed date.
+  </span>
+  <span *ngIf="reactiveDate.errors.required">
+    Date is required. Enter a value.
```
Comment (blackbaud-johnly) [inline src/demos/datepicker/fuzzy-datepicker-demo.component.html]: This is missing "The" or "A" at the beginning. Also, do we need the second sentence here? I don't think so, so I'd delete it.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 10 above)
Comment (blackbaud-johnly) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Actually, thinking on this more, the error message is also vague. What makes a date valid? Can we change this to `Please enter a date in the [required date format goes here] format.`? (I know that flipflops on the enter/select thing, but now I realize that if you "select" the date, then of course it's going to be in the correct format. Sorry about that.)

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+  </span>
+</p>
+
+<p>
+  <button
+    class="sky-btn sky-btn-secondary"
+    (click)="clearSelectedDates()"
+  >
+    Clear selected dates
```
Comment (blackbaud-johnly) [inline src/demos/datepicker/fuzzy-datepicker-demo.component.html]: It's superminor, bit It'd delete "selected." It doesn't add anything; it just makes the button label longer.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 2 above)
Comment (blackbaud-johnly) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: If we leave it as is, I'd change "Set" to "Enable" since the user isn't really setting min and max dates, they are just turning them on.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+    (click)="setMinMaxDates()"
+  >
+    Set min and max dates
+  </button>
+  <button
+    class="sky-btn"
+    type="button"
+    (click)="toggleYearRequired()"
+  >
+    Toggle year required
```
Comment (blackbaud-johnly) [inline src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Would it be difficult to have 2 versions of this label? Like maybe `Enable year requirement` and `Disable year requirement`? (If it's difficulty, it's probably not worth it. If not, let me know, and I can try to massage those labels a bit more.)

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+  </button>
+  <button
+    class="sky-btn"
+    type="button"
+    (click)="togglefutureDisabled()"
+  >
+    Toggle cannot be in cannot be in future
```
Comment (blackbaud-johnly) [inline src/demos/datepicker/fuzzy-datepicker-demo.component.html]: We've got duplicate "cannot be in" instances that make this label gibberish. ... But even if we fix that, this is a vague label. I'd change this to something like `Toggle ability to add future dates` or `Toggle future date restriction`. (That said, if we can have 2 versions of this label, like I asked about for the previous button, we could tighten this up and make it clearer: `Enable future dates` and `Disable future dates`.)

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html [hunk 15]
```diff
@@ -0,0 +1,133 @@
+<h2
+  class="sky-section-heading"
+>
+  Standard fuzzy datepicker
```
Comment (blackbaud-johnly) [inline src/demos/datepicker/fuzzy-datepicker-demo.component.html]: What is "standard" about this? Using that terminology makes it sound like "recommended," which I don't think is what we mean to convey. Should "Standard" be "Template-driven"? Or something else?

File: src/demos/demo.service.ts [hunk 16]
```diff
@@ -431,6 +432,22 @@ export class SkyDemoService {
         }
       ]
     },
+    {
+      name: 'Fuzzy datepicker',
```
Comment (blackbaud-johnly) [inline src/demos/demo.service.ts:436]: Should this really be a separate entry from the datepicker demo?

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 2 above)
Comment (Blackbaud-AlexKingman) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Added toggles.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 1 above)
Comment (Blackbaud-AlexKingman) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Updated to toggles. Added spacing.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 3 above)
Comment (Blackbaud-AlexKingman) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Done.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 4 above)
Comment (Blackbaud-AlexKingman) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Done.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 5 above)
Comment (Blackbaud-AlexKingman) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Done.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 6 above)
Comment (Blackbaud-AlexKingman) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Done.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 8 above)
Comment (Blackbaud-AlexKingman) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Done.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 10 above)
Comment (Blackbaud-AlexKingman) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Done.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 11 above)
Comment (Blackbaud-AlexKingman) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Done.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 12 above)
Comment (Blackbaud-AlexKingman) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Changed to "Reset dates".

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 13 above)
Comment (Blackbaud-AlexKingman) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Updated to toggle switches.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 14 above)
Comment (Blackbaud-AlexKingman) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Updated to toggle switches.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 15 above)
Comment (Blackbaud-AlexKingman) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Changed to "Template-driven".

File: src/demos/demo.service.ts (hunk 16 above)
Comment (Blackbaud-AlexKingman) [reply src/demos/demo.service.ts:436]: Yeah, we actually have a pattern of this with other demos (filters). Keeping the separate helps us isolate changes and keep these modular.

**Review by Blackbaud-BobbyEarl — DISMISSED** [review]

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 4 above)
Comment (blackbaud-johnly) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: In an actual client-facing error message, we wouldn't display the date as `{"month":#,"day":#,"year":####}`. Does that matter here? I know this and the error message for dates before the threshold are just demo examples, but if we format it that way in our demos that way, are we implying to developers that this is appropriate? How much work would it be to format the date instead of just displaying the object/array?

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 5 above)
Comment (blackbaud-johnly) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Same question here about formatting the date. I should have thought of this yesterday. Sorry. ... It may not be an issue, but it seems worth asking the question. Maybe we need to get Todd to weigh in on whether this is worth addressing?

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 6 above)
Comment (blackbaud-johnly) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Ditto previous messages on date format.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 3 above)
Comment (blackbaud-johnly) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: See question that follows on date format.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 7 above)
Comment (blackbaud-johnly) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Looks good, but same question about formatting the date.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 8 above)
Comment (blackbaud-johnly) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Looks good, but same question about formatting the date.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 9 above)
Comment (blackbaud-johnly) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Looks like this didn't change. Was that intentional?

File: src/demos/demo.service.ts (hunk 16 above)
Comment (blackbaud-johnly) [reply src/demos/demo.service.ts:436]: Good deal. I wasn't sure. I'll keep that in mind in the future. Thanks!

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 4 above)
Comment (Blackbaud-AlexKingman) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Yep - I agree. Thanks for pushing me on that. 🏆

Comment (Blackbaud-AlexKingman) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Done. Date formatted.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 5 above)
Comment (Blackbaud-AlexKingman) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: This part is meant to be for the developers. It's showing you what the object looks like. It's not intended for end users to see. Should we make that more clear?

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 6 above)
Comment (Blackbaud-AlexKingman) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: See previous comment (re: object data).

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 7 above)
Comment (Blackbaud-AlexKingman) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Date formatted.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 8 above)
Comment (Blackbaud-AlexKingman) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Date formatted.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 9 above)
Comment (Blackbaud-AlexKingman) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Nope, it fell through the cracks. Good catch! Fixed.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 3 above)
Comment (Blackbaud-AlexKingman) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: Date formatted.

File: src/demos/datepicker/fuzzy-datepicker-demo.component.html (hunk 5 above)
Comment (blackbaud-johnly) [reply src/demos/datepicker/fuzzy-datepicker-demo.component.html]: I can't think of a way to make that more clear. Let's leave it as is for now, and I'll circle back if I can think of a better way to make this distinction.

**Review by blackbaud-johnly — APPROVED** [review]
LGTM
